util/UnitEvent.lua

```diff
@@ -0,0 +1,165 @@
+
+function UnitEvent.registerCallback(eventType, func)
+    if eventType > 2 or eventType < 1 then return end
+    table.insert(trigTable[eventType + 2], func)
+end
+
+function UnitEvent.getEventType() return evType end
+function UnitEvent.getTriggerUnit() return evData end
+function UnitEvent.getTriggerUnitId() return GetHandleId(evData) end
+
+local function callback()
+    local i = evType + 2
+    for _, func in pairs(trigTable[i]) do
+        execute(func)
+    end
+end
+
+ceres.addHook("main::before", function()
+    TriggerAddCondition(trigTable[0], Filter(function()
+        local issuedOrder   = GetIssuedOrderId()
+        local u             = GetTriggerUnit()
+        --  Magic undefense
+        if issuedOrder == 852479 then
+            if BlzGetUnitAbility(u, DETECT_LEAVE) == nil then
+                local i     = GetUnitId(u)
+                if GetUnitTypeId(u) == 0 then
+                    userData.indexed[i]         = nil
+                    userData[i]                 = nil
+                    userData.preplaced[i]       = nil
+                    userData.unitId[i]          = nil
+                    userData.nativeHasAbil[i]   = nil
+
+                    --  Only do callback if the game has already initialized.
+                    if initialized then
+                        local l     = evType
+                        local ij    = evData
+                        
+                        evType      = EVENT_UNIT_LEAVE
+                        evData      = u
+                        callback()
+                        evData      = ij
+                        evType      = l        
+                    end
+                end
+            end
+        end
+    end))
+    for i=0, bj_MAX_PLAYER_SLOTS - 1 do
+        SetPlayerAbilityAvailable(players[i], DETECT_LEAVE, false)
+        TriggerRegisterPlayerUnitEvent(trigTable[0], players[i], EVENT_PLAYER_UNIT_ISSUED_ORDER, nil)
+    end
+ 
+    TriggerRegisterEnterRegion(trigTable[1], World.REG, nil)
+    TriggerAddCondition(trigTable[1], Filter(function()
+        local u = GetTriggerUnit()
+        local b = UnitAddAbility(u, DETECT_LEAVE)
+        local i = GetUnitId(u)
+
+        UnitMakeAbilityPermanent(u, true, DETECT_LEAVE)
+        if not userData.indexed[i] then
+            userData.indexed[i]         = true
+            userData.preplaced[i]       = not initialized
+            userData.unitId[i]          = u
+            userData.nativeHasAbil[i]   = not b
+            
+            local l     = evType
+            local ij    = evData
+
+            if initialized then
+                evType      = EVENT_UNIT_ENTER
+                evData      = u
+                callback()
+                evData      = ij
+                evType      = l
+            end
+        end
+    end))
+end)
+
+ceres.addHook('main::after', function()
+    GroupEnumUnitsInRect(trigTable[2], World.RECT, nil)

```

Comment:
If I recall correctly, there's no way to have units created before the game starts without `CreateUnit` native, so I suggest replacing the native.

Reply:
Considering that, I'd have to replace CreateCorpse as well.